### ISSUE TITLE
Spike: Install the API server on the cluster

### ISSUE THREAD

**Issue by jpellizzari** [weaveworks/weave-gitops#499]
We want the UI (and therefore) the API server to be a resident on the cluster. This issue seeks to enumerate the design decisions before starting on implementation.

Open questions:

* What is the network path for accessing the API? Do we utilize kubeproxy?

* We will need a docker container with our `wego-server` binary running. How does that get created and stored? For development, can we add an option to build the container from source instead of pulling it from an image repo?

* A service account will need to be created for the API to use. What permissions does that account have?

**Comment by josecordaz** [comment]
Proposal for **api-service** docker image production path. 

We had some github actions that are related to docker images (build, analysis for security issues) so might be able to rescue that and apply it to the **api-service** image publish process. It would be triggered on tag and PRs events. Tempted actions would be:

1.- Build the docker image
2.- Analyze docker image for for security issues
3.- Push it to dockerhub registry using a combination of release tag and the short commit to create the docker tag. This would work for all scenarios (release and PRs). We can't use only the tag because the LD value for tag is always `v0.0.0` for binaries created on PRs events. For this particular case if we combine with the short commit it would create a docker tag with a value of `v0.0.0-1200b7f`(for instance) so when this particular binary runs, it will use that specific tag to update the image in the deployment file at runtime. For development we can prioritize an environment variable with the desired api-service image tag (in case we want to use a different one in combination with a tool that allows the use of local docker images). If there is a value in the env we use that if not we use the LD values for tag and short commit.
- IMPORTANT NOTE: We could cache the context of the build as well so the builds could be fast.

**Comment by jpellizzari** [comment]
> CONCLUSION. I think using the kubectl proxy is a good idea. I spoke with with Simon, and seems they a have a story to switch to kubectl proxy as well.

How does that work? Can you list out the steps that need that the code needs to take?

>  Push it to dockerhub registry using a combination of release tag and the short commit to create the docker tag. This would work for all scenarios (release and PRs)

Why do we need to push on PR events? I am worried this will clutter up the registry with lots of non-usable docker images. Can we push only on releases instead?

> Push it to dockerhub registry

Can we use [GitHub Container Registry](https://docs.github.com/en/packages/working-with-a-github-packages-registry/working-with-the-container-registry) instead?

> There is a couple of places in Add where we are getting resources from different namespaces other than wego-system. Not sure we we want to keep them there

What are the consequences/tradeoffs for moving all of the flux objects into the `wego-system` namespace?

**Comment by josecordaz** [comment]
@jpellizzari 

> How does that work? Can you list out the steps that need that the code needs to take?

1.- Run `kubectl proxy`. It usually runs on `127.0.0.1:8001`

2.- Create the api-service path using this formula `http://HOST:PORT/api/v1/namespaces/NAMESPACE/services/SERVICE-NAME/proxy/`
Replace HOST with `127.0.0.1`
Replace PORT with `8001`
Replace NAMESPACE with `wego-system` (or the NS we decide to install our service resource on)
Replace SERVICE-NAME with `simple-service` (or the service-name we decide to identify our service resource with)

Docs on how to build the URL here: https://kubernetes.io/docs/tasks/administer-cluster/access-cluster-services/#manually-constructing-apiserver-proxy-urls

>  Why do we need to push on PR events? I am worried this will clutter up the registry with lots of non-usable docker images. Can we push only on releases instead?

I'm thinking this for cases where we work specifically on changes to the api-service, we probably want to generate an image directly from the changes made in the PR, right now I don't think there is a way to know if the PR contains api-service related changes or not but totally open to suggestions. We certainly can push only in releases but the only case I'm worried about is on changes to the api-service because we might want to generate an image with those changes so the test suites can use it to verify functionality. I'm not 100% sure there will be non-usable docker images, my assumption is that the tests are going to use the one created in its PR but I could be wrong. There might be docker images that never reach the main branch. But that' is where a shared docker build context can help up us out to speed up the builds even for releases.

> Can we use GitHub Container Registry instead? One less external system for us to authenticate with. 

I'm totally ok with this.

> What are the consequences/tradeoffs for moving all of the flux objects into the wego-system namespace?

I think we already have all flux objects on that namespace. In the past I mentioned that by using Role and RoleBinding we might face some issues because applications might have resources across different namespaces and such. But I forgot that flux is the one who actually applies those manifests in those paths. So probably this is already covered.

**Comment by jpellizzari** [comment]
>  I'm worried about is on changes to the api-service because we might want to generate an image with those changes so the test suites can use it to verify functionality. 

The best practice would be to always build images from source, rather than pulling in an artifact. Like you mentioned, the cached build context and image layers will be key for ensuring the tests build images quickly.

We will always build a new container image on every release and tag it with the git tag. Let's also make sure we utilize two-stage builds where possible, so that the runtime container has a very minimal surface area.

LGTM :+1:

**Comment by josecordaz** [comment]
> We will always build a new container image on every release and tag it with the git tag. Let's also make sure we utilize two-stage builds where possible, so that the runtime container has a very minimal surface area.

Right, 100% agree. 🙂

**Comment by JamWils** [comment]
A lot of this seems important, should we document these decisions in ADR(s)?  The api-server in particular.

**Comment by jpellizzari** [comment]
> @josecordaz and @jpellizzari, a lot of this seems important, should we document these decisions in ADR(s)? The api-server in particular.

It's probably big enough to warrant that. Let's plan on doing that, since these comments are hard to follow.

**Comment by JamWils** [comment]
@josecordaz with the ADR merged is there anything left for this spike or can we close it out?

**Comment by josecordaz** [comment]
@JamWils we can close it 🙂
